design/Extend-VolumePolicies-to-support-more-actions.md

```diff
@@ -0,0 +1,161 @@
+# Extend VolumePolicies to support more actions
+
+## Abstract
+
+Currently, the [VolumePolicies feature](https://github.com/vmware-tanzu/velero/blob/main/design/Implemented/handle-backup-of-volumes-by-resources-filters.md) which can be used to filter/handle volumes during backup only supports the skip action on matching conditions. Users need more actions to be supported.
+
+## Background
+
+The `VolumePolicies` feature was introduced in Velero 1.11 as a flexible way to handle volumes. The main agenda of introducing the VolumePolicies feature was to improve the overall user experience when performing backup operations for volume resources, the feature enables users to group volumes according the `conditions` (criteria) specified and also lets you specify the `action` that velero needs to take for these grouped volumes during the backup operation. The limitation being that currently `VolumePolicies` only supports `skip` as an action, We want to extend the `action` functionality to support more usable options like `fs-backup-volumes` (File system backup) and `fs-backup-volumes-excludes` (VolumeSnapshots).
+
+## Goals
+- Extending the VolumePolicies to support more actions like file system backup(`fs-backup-volumes`) and snapshots(`fs-backup-volumes-excludes`).
+- Improve user experience when backing up Volumes via Velero
+
+## Non-Goals
+- No changes to existing approaches to opt-in/opt-out annotations for volumes
+- No changes to existing `VolumePolicies` functionalities
+
+
+## Use-cases/Scenarios
+
+**Use-case 1:**
+- A user wants to use `fs-backup-volumes-excludes` (volumesnapshots) backup option for all the csi supported volumes and `fs-backup-volumes` for the rest of the volumes.
+- Currently, velero supports this use-case but the user experience is not that great.
+- The user will have to individually annotate the volume mounting pod with the annotation "backup.velero.io/backup-volumes" for `fs-backup-volumes`
+- This becomes cumbersome at scale.
+- Using `VolumePolicies`, the user can just specify a simple `VolumePolicy` like for csi supported volumes and rest can be defaulted to `fs-backup-volumes`:
+```yaml
+version: v1
+volumePolicies:
+- conditions:
+    csi:
+      driver: aws.efs.csi.driver
+  action:
+    type: fs-backup-volumes-excludes
+```
+
+**Use-case 2:**
+- A user wants to use `fs-backup-volumes` for nfs volumes
+- In such a scenario the user can just specify a `VolumePolicy` like:
+```yaml
+version: v1
+volumePolicies:
+- conditions:
+    nfs:
+      server: 192.168.200.90
+  action:
+    type: fs-backup-volumes
+```
+## High-Level Design
```

Comment:
Should we consider adjusting the backup workflow? In our current backup process, the procedures for `fs-backup`, `csi`, and `volume snapshot` are all mixed together. Should we consider clarifying these different backup workflows for easier maintenance in the future?

Reply:
@qiuming-best I'm thinking that backup workflow redesign may be a separate thread here. We may well want to do it, but this particular policy change isn't adding any new workflow steps, it's just providing an alternate means for users to specify volumes for fs-backup opt-in and opt-out, so using the current workflow may be the clearest. Then in a separate development task, we can redesign the workflow if there's a need.

Reply:
@qiuming-best IMO Redesigning backup workflow  is not scope of this design, we need to have a different design for that, The scope of this design is to improve the UX of already existing functionalities for the user, re-iterating what @sseago said here, enabling users with a better option to the already existing annotation based opt-in/opt-out approach is the exact goal and use-case of this design.

Reply:
> Enabling users with a better option to the already existing annotation-based opt-in/opt-out approach

This is the root of the disagreement here...  IMO the current opt-in/out design, when combined with other options, has created a lot confusions and corner cases, and we don't wanna repeat that.

Reply:
Resolving this comment as we agreed not to copy the opt-in/opt-out approach into volumepolicies.